### PR TITLE
Extract UI from Rails

### PR DESCRIPTION
**Pull request by Connoropolous**

### CODE REVIEW

File: .codeclimate.yml [hunk 1]
```diff
@@ -1,38 +1,22 @@
 ---
 engines:
-  brakeman:
```
Comment (devvmh) [inline .codeclimate.yml:3]: OK, so I was confused by this a bit, but I see now -> you're not gonna merge this PR, but rather create a new JS-only repo with it, which won't include ruby. Hence it won't need rspec/brakeman. right?

File: LICENSE [hunk 2]
```diff
@@ -1,661 +0,0 @@
-GNU AFFERO GENERAL PUBLIC LICENSE
```
Comment (devvmh) [inline LICENSE:1]: should we still have a license in the new repo?

File: README.md [hunk 3]
```diff
@@ -1,62 +1,82 @@
...
-[license]: https://github.com/metamaps/metamaps/blob/develop/LICENSE
-[contributing]: https://github.com/metamaps/metamaps/blob/develop/doc/CONTRIBUTING.md
-[contributing-issues]: https://github.com/metamaps/metamaps/blob/develop/doc/CONTRIBUTING.md#reporting-bugs-and-other-issues
-[mac-installation]: https://github.com/metamaps/metamaps/blob/develop/doc/MacInstallation.md
-[ubuntu-installation]: https://github.com/metamaps/metamaps/blob/develop/doc/UbuntuInstallation.md
-[vagrant-installation]: https://github.com/metamaps/metamaps/blob/develop/doc/VagrantInstallation.md
-[windows-installation]: https://github.com/metamaps/metamaps/blob/develop/doc/WindowsInstallation.md
+[![Build Status](https://travis-ci.org/metamaps/metamaps-ui.svg?branch=master)](https://travis-ci.org/metamaps/metamaps-ui)
+
+Make sure you're running a good up to date LTS version of `node`, like 8.9.4
```
Comment (devvmh) [inline README.md]: suggest to add "at time of writing"

File: README.md [hunk 4]
```diff
@@ -1,62 +1,82 @@
...
+
+Make sure you have `node-sass` installed
+`$ npm install -g node-sass`
+
+Run the following at the same time, in TWO SEPARATE terminals. We tell the server where the backend process is running with the API environment variable
+JS files, and CSS will rebuild automatically, just refresh the page
+If coding the server itself, you will have to use nodemon, or kill and restart the server process manually
+```
+$ API=http://localhost:3001 node server.js
+$ node-sass -w sass/application.scss public/css/application.css
```
Comment (devvmh) [inline README.md]: the nodemon stuff  + two separate terminals is hard to follow. Want me to try drafting an iteration? or do you want to take a 2nd pass first?

File: README.md [hunk 5]
```diff
@@ -1,62 +1,82 @@
...
+touch sass/application.scss
+```
+
+Run the metamaps api in another terminal using (on port 3001, so the UI can talk to it)
+For now, make sure you are running on the `add-user-route` branch of Metamaps, and that it's up to date with the latest on that branch
+`$ rails s -p 3001`
+
+open up http://localhost:3000 and start coding!
+
+Checklist
```
Comment (devvmh) [inline README.md:34]: is this still up to date? it will be removed before merge, right?

File: README.md [hunk 6]
```diff
@@ -1,62 +1,82 @@
...
+$ API=http://localhost:3001 node server.js
+$ node-sass -w sass/application.scss public/css/application.css
+```
+
+To make sure the css files get built, use the following in another terminal
+```
+touch sass/application.scss
+```
+
+Run the metamaps api in another terminal using (on port 3001, so the UI can talk to it)
```
Comment (devvmh) [inline README.md]: so will we be splitting this into two repos?

1. metamaps-ui
2. metamaps-api

and both would contain the full  history of metamaps, with a commit in each that demarcates where they were split?

TBH I'd prefer to keep them in one repo if possible (with a nice README/script for running all the servers). if it's not possible, it would be cool to keep the history in both repos and cross link

File: apiProxyMiddleware.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    if (!(req.xhr || req.headers['content-type'] === 'application/json' || req.originalUrl.indexOf('.json') > -1 || req.method !== 'GET')) {
+        return next()
+    }
+    const method = req.method.toLowerCase()
+    req.pipe(
+        request[method](process.env.API + req.originalUrl, {
+            headers: {
+                ...req.headers,
+                cookie: `_Metamaps_session=${req.cookies._Metamaps_session}`,
+                host: 'localhost:3001'
```
Comment (devvmh) [inline apiProxyMiddleware.js]: make port configurable with ENV or something

File: apiProxyMiddleware.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+const request = require('request')
+
+function apiProxyMiddleware (req, res, next) {
+    // TODO: tidy this up!
```
Comment (devvmh) [inline apiProxyMiddleware.js]: would be nice to have a header with a description of the purpose/usage, since it's a top level file

File: app/assets/javascripts/homepageVimeoFallback.js.erb [hunk 9]
```diff
@@ -1,29 +0,0 @@
-/* global $ */
```
Comment (devvmh) [inline app/assets/javascripts/homepageVimeoFallback.js.erb:1]: this will stay in the ruby repo, right?

will there be any other JS in the ruby repo?

File: package.json [hunk 10]
```diff
@@ -20,6 +22,7 @@
   },
   "homepage": "https://github.com/metamaps/metamaps#readme",
   "dependencies": {
+    "action-cable-node": "Connoropolous/action-cable-node",
```
Comment (devvmh) [inline package.json:26]: !! what's going on here. do you have a PR open to get this change merged upstream?

File: app/assets/javascripts/homepageVimeoFallback.js.erb (hunk 9 above)
Comment (devvmh) [reply app/assets/javascripts/homepageVimeoFallback.js.erb:1]: i see the video thumbnail is still here, maybe we do need this JS after all. consider it

File: public/index.html [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+      { elsif notice }
+        Metamaps.ServerData.toast = "{ notice }"
+      { elsif alert }
+        Metamaps.ServerData.toast = "{ alert }"
+      { end }
+
+      { render :partial => 'shared/metacodeCssColors' }
+      { render :partial => 'layouts/googleanalytics' if ENV["GA_TRACKING_CODE"].present? }
+
+    -->
```
Comment (devvmh) [inline public/index.html:104]: i guess this stuff can be deleted? was this in the main repo? we should probably delete if so

File: public/index.html [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+  <script type="text/javascript">
+    Metamaps.ServerData = Metamaps.ServerData || {}
+    Metamaps.ServerData.RAILS_ENV = 'development'
```
Comment (devvmh) [inline public/index.html]: this will need to be changed

File: public/index.html [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+    Metamaps.ServerData['synapse16.png'] = '/images/synapse16.png'
+    Metamaps.ServerData['sounds/MM_sounds.mp3'] = '/images/sounds/MM_sounds.mp3'
+    Metamaps.ServerData['sounds/MM_sounds.ogg'] = '/images/sounds/MM_sounds.ogg'
+    Metamaps.ServerData['exploremaps_sprite.png'] = '/images/exploremaps_sprite.png'
+    Metamaps.ServerData['map_control_sprite.png'] = '/images/map_control_sprite.png'
+    Metamaps.ServerData['user_sprite.png'] = '/images/user_sprite.png'
+    Metamaps.Loading.setup()
+  </script>
+
+  <div class="templates">
```
Comment (devvmh) [inline public/index.html:127]: these templates should all become react components

if it makes it easier, we could even have JS to render the. templates given ServerData right here calling into the components

File: public/lib/jquery-1.12.4.min.js [hunk 14]
```diff
@@ -0,0 +1,5 @@
+/*! jQuery v1.12.4 | (c) jQuery Foundation | jquery.org/license */
```
Comment (devvmh) [inline public/lib/jquery-1.12.4.min.js:1]: can we pull this via npm now?

Comment (devvmh) [reply public/lib/jquery-1.12.4.min.js:1]: same with jquery-ui

File: realtime/README.md [hunk 15]
```diff
@@ -1,17 +0,0 @@
-## Node.js realtime server
```
Comment (devvmh) [inline realtime/README.md:1]: should we have a new realtime README

File: sass/application.scss [hunk 16]
```diff
@@ -1,17 +1,3 @@
-/*
```
Comment (devvmh) [inline sass/application.scss:1]: on the sass directory - why sass? are we doing all sass, no css? it could be called 'stylesheets' if you wanted to be flexible?

i guess sass does make sense though since css is a subset of sass

File: src/Metamaps/GlobalUI/Notifications.js [hunk 17]
```diff
@@ -54,7 +83,7 @@ const Notifications = {
         }
       },
       error: function() {
-        GlobalUI.notifyUser('There was an error marking that notification as read')
+        GlobalUI.notifyUser('There was an error marking that notification as unread')
```
Comment (devvmh) [inline src/Metamaps/GlobalUI/Notifications.js:86]: was this a bug? why are there not errors for both cases?

File: src/Metamaps/GlobalUI/ReactApp.js [hunk 18]
```diff
@@ -93,7 +91,7 @@ const ReactApp = {
     const self = ReactApp
     const createElement = (Component, props) => <Component {...props} {...self.getProps()}/>
     const app = <Router createElement={createElement} routes={routes} history={browserHistory} onUpdate={self.handleUpdate} />
-    ReactDOM.render(app, document.getElementById('react-app'))
+    ReactDOM.render(app, document.getElementById('app'))
```
Comment (devvmh) [inline src/Metamaps/GlobalUI/ReactApp.js:94]: maybe we should have a more specific selector than #app, just in case this is rendered alongside other apps or something now that's it's extractable. what about #metamaps-app?

File: src/Metamaps/GlobalUI/index.js [hunk 19]
```diff
@@ -137,19 +140,6 @@ const GlobalUI = {
       ReactApp.render()
       self.notifying = false
     }
-  },
-  shareInvite: function(inviteLink) {
```
Comment (devvmh) [inline src/Metamaps/GlobalUI/index.js:141]: what happened to shareInvite? is it just broken? I see clipboard js is still a. dependency

File: src/Metamaps/index.js [hunk 20]
```diff
@@ -78,9 +75,31 @@ document.addEventListener('DOMContentLoaded', function() {
...
       typeof (Metamaps[prop].init) === 'function'
     ) {
-      Metamaps[prop].init(Metamaps.ServerData)
+      Metamaps[prop].init(serverData)
     }
   }
+}
+
+// fetch data from API then pass into init functions
+document.addEventListener('DOMContentLoaded', async function() {
```
Comment (devvmh) [inline src/Metamaps/index.js:84]: i think this is our only async abstraction right? other than jquery promises? it's nice to standardize on one where possible (I.e. no sagas, etc)

Comment (devvmh) [reply src/Metamaps/index.js:84]: FWIW i prefer native es6 promises to async/await where possible though, because of their explicitness

File: src/components/LightBoxes/ForkMap.js [hunk 21]
```diff
@@ -0,0 +1,66 @@
+import React, { Component } from 'react'
+
+class ForkMap extends Component {
```
Comment (devvmh) [inline src/components/LightBoxes/ForkMap.js:3]: i don't want to make too much extra work, but this should really be split into smaller subcomponents e.g. permIcon, clearFloat, and maybe input group components

File: src/components/LightBoxes/Invite.js [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+        <p>Below is a personal invite link containing your unique access code, which can be used multiple times.</p>
+        <div id="joinCodesBox">
+          <p className="joinCodes">
+            {inviteLink}
+            <button className="button" onClick={this.shareInvite}>COPY INVITE LINK!</button>
+          </p>
+          <p className="popup" style={{textAlign: 'center'}}>
+            {this.state.copied && 'Copied!'}
+            {this.state.unable && "Your browser doesn't support copying, please copy manually."}
+          </p>
```
Comment (devvmh) [inline src/components/LightBoxes/Invite.js:49]: arguably these should be two different paragraphs (each rendered within one of the conditions), because if both were true for some unforeseen reason then we'd see a weird  sentence with "Copied!" tacked on the front

File: src/components/LightBoxes/NoIE.js [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+  render = () => {
+    return (
+      <div className="lightboxContent" id="noIE">
+        <h3>OOPS! <br /> YOUR BROWSER IS NOT SUPPORTED.</h3>
+        <p id="noIEsubheading">To view this experience, please upgrade to the latest one of these browsers:</p>
+        <a id="chromeIcon" href="https://www.google.com/chrome/browser/" target="_blank">Chrome</a>
+        <a id="fireFoxIcon" href="https://www.mozilla.org/en-US/firefox/new/" target="_blank">Firefox</a>
+        <a id="safariIcon" href="http://support.apple.com/downloads/#safari" target="_blank">Safari</a>
+        <p id="noIEbody">
+          While it's downloading, explore our <a href="http://blog.metamaps.cc/">blog</a>,<br />
```
Comment (devvmh) [inline src/components/LightBoxes/NoIE.js:13]: remove the blog link since it's no longer up

**Review by devvmh — COMMENTED** [review]
first chunk of comments. I got as far as src/components/LightBoxes/SwitchMetacodes.js

File: .codeclimate.yml (hunk 1 above)
Comment (Connoropolous) [reply .codeclimate.yml:3]: right!

File: LICENSE (hunk 2 above)
Comment (Connoropolous) [reply LICENSE:1]: yeah

File: README.md (hunk 3 above)
Comment (Connoropolous) [reply README.md]: the readme will need lots of updates, noted about this one

File: README.md (hunk 4 above)
Comment (Connoropolous) [reply README.md]: I'll try again with it

File: README.md (hunk 5 above)
Comment (Connoropolous) [reply README.md:34]: it's up to date with what I've accomplished so far, but yeah it'll be removed. And ultimately, I'm aiming for the only thing that will get merged into metamaps main repo are the ones necessary to make it standalone from the UI

File: README.md (hunk 6 above)
Comment (Connoropolous) [reply README.md]: yeah, I agree with keeping the history, but I think going with two repos is the way, since what we are starting to envision is the possible evolution away from the rails backend

File: apiProxyMiddleware.js (hunk 7 above)
Comment (Connoropolous) [reply apiProxyMiddleware.js]: 👍

File: apiProxyMiddleware.js (hunk 8 above)
Comment (Connoropolous) [reply apiProxyMiddleware.js]: yeah

File: package.json (hunk 10 above)
Comment (Connoropolous) [reply package.json:26]: the main repo itself seems pretty sketchy... since there was a TYPO! in the code, but I was looking for a solution, and this was one I found. I do have a PR in to get it fixed, but haven't heard back

File: public/index.html (hunk 11 above)
Comment (Connoropolous) [reply public/index.html:104]: yeah... there's lots I'm still aiming to delete, I have comments still sitting where there's pieces of work left to consider

File: public/index.html (hunk 13 above)
Comment (Connoropolous) [reply public/index.html:127]: The mapinfobox can become a react component, I've been seeing if @Bortseb will take on some of that, but the rest of them which have to do with twitter autocomplete library can't, unless we move to a react autocomplete library

File: public/lib/jquery-1.12.4.min.js (hunk 14 above)
Comment (Connoropolous) [reply public/lib/jquery-1.12.4.min.js:1]: yeah

File: realtime/README.md (hunk 15 above)
Comment (Connoropolous) [reply realtime/README.md:1]: we don't really need one, since the realtime server is served through the same server as this whole js repo, we could make some comments in the readme about it

File: sass/application.scss (hunk 16 above)
Comment (Connoropolous) [reply sass/application.scss:1]: yeah, if we start doing css right, it should prbly all be sass, so setting our sights on that

File: src/Metamaps/GlobalUI/Notifications.js (hunk 17 above)
Comment (Connoropolous) [reply src/Metamaps/GlobalUI/Notifications.js:86]: there are errors for both cases I think, just the error message for the marking unread one was wrong

File: src/Metamaps/GlobalUI/ReactApp.js (hunk 18 above)
Comment (Connoropolous) [reply src/Metamaps/GlobalUI/ReactApp.js:94]: sounds good

File: src/Metamaps/GlobalUI/index.js (hunk 19 above)
Comment (Connoropolous) [reply src/Metamaps/GlobalUI/index.js:141]: the functionality got moved into the Invite.js lightbox, it's all working

File: src/Metamaps/index.js (hunk 20 above)
Comment (Connoropolous) [reply src/Metamaps/index.js:84]: yeah, I've been exploring the async/await syntax, and actually, I really like it, but I'm not attached to it. The whole `DataFetcher.js` file that I created is using it too though, and scattered in a few of the react class methods which make async calls

File: src/components/LightBoxes/ForkMap.js (hunk 21 above)
Comment (Connoropolous) [reply src/components/LightBoxes/ForkMap.js:3]: Hmm, I could see about doing that.

File: src/components/LightBoxes/Invite.js (hunk 22 above)
Comment (Connoropolous) [reply src/components/LightBoxes/Invite.js:49]: it should definitely be one or the other, because of the clipboardjs api, but I guess we can reflect a better style in the component, just in case that other bit changes in the future

File: src/components/LightBoxes/NoIE.js (hunk 23 above)
Comment (Connoropolous) [reply src/components/LightBoxes/NoIE.js:13]: 👍

File: src/routes/UserSettings.js [hunk 24]
```diff
@@ -156,17 +142,65 @@ class UserSettings extends Component {
     }
   }
 
+  onSubmit = async (event) => {
+    event.preventDefault()
+    const { updateUser, params: { id } } = this.props
+    const { name, email, imagePreviewFile, emailsAllowed, removeImage,
+            followTopicOnCreated, followTopicOnContributed,
+            followMapOnCreated, followMapOnContributed,
+            currentPassword, newPassword, newPasswordConfirmation  } = this.state
```
Comment (devvmh) [inline src/routes/UserSettings.js:151]: if the state is this big, this should be refactored into smaller subcomponents

File: src/routes/UserSettings.js [hunk 25]
```diff
@@ -156,17 +142,65 @@ class UserSettings extends Component {
...
+    this.setState({ loading: true, error: false })
+    try {
+      const res = await updateUser(id, {
+        name,
+        email,
+        removeImage,
```
Comment (devvmh) [inline src/routes/UserSettings.js:157]: `removeImage` is a confusing name, what about `showRemovalImage`?

File: src/Metamaps/DataFetcher.js [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+      'Content-Type': 'application/json'
+    }
+  })
+}
+
+async function getMetacodes() {
+  const res = await get('/metacodes')
+  const data = await res.json()
+  return data
+}
```
Comment (devvmh) [inline src/Metamaps/DataFetcher.js:62]: I'm fine with this file as-is but I'd prefer to unify the code on Promises or async/await, and I think promises are more flexible.  I'm not sure if it works this way, but it would be cool to use async as an annotation that indicates a function returns a promise. So I'd prefer

```
async function getMetacodes() {
  return get('/metacodes')
    .then(response => response.json())
}
```

I'm not too dogmatic about this though - if you like the look of async/await better, let's try it out

File: src/Metamaps/DataFetcher.js [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+
+function deleteReq(url) {
+  return fetch(url, {
+    credentials: 'same-origin',
+    method: 'DELETE',
+    headers: {
+      'Content-Type': 'application/json'
+    }
+  })
+}
```
Comment (devvmh) [inline src/Metamaps/DataFetcher.js:56]: suggest defining these in terms of a common helper function

```
function commonFetch(url, method, data = {}, preProcessBody = JSON.stringify) {
  return fetch(url, {
    credentials: 'same-origin',
    headers: {
      'Content-Type': 'application/json'
    },
    method, 
    body: data ? preProcessBody(data) : undef
  })
}

function deleteReq(url) {
  return commonFetch(url, 'DELETE', undef)
}

// ...etc
```

**Review by devvmh — COMMENTED** [review]
DataFetcher comments

File: src/components/LightBoxes/Invite.js (hunk 22 above)
Comment (devvmh) [reply src/components/LightBoxes/Invite.js:49]: yeah, like 

```
{this.state.copied && <Popup message="Copied!" />
...
const Popup = ({ message }) => {
  return <p className="popup" style={{textAlign: 'center'}}>
    {message}
  </p>
}
```

File: src/Metamaps/index.js (hunk 20 above)
Comment (devvmh) [reply src/Metamaps/index.js:84]: i commented more below. I definitely like async as a function annotation to indicate that the function returns a promise. I think promises are a bit more explicit. I'm not sure where I'll land on this though

File: package.json (hunk 10 above)
Comment (devvmh) [reply package.json:26]: damn

Comment (devvmh) [reply package.json:26]: if you do hear back, let's ask to be collaborators on github/npm so we can push fixes

File: public/lib/jquery-1.12.4.min.js (hunk 14 above)
Comment (devvmh) [reply public/lib/jquery-1.12.4.min.js:1]: awesome! this is a really great change. It's a big step to moving off of jquery entirely if we want someday, which could help get our bundle size down

File: sass/application.scss (hunk 16 above)
Comment (devvmh) [reply sass/application.scss:1]: I'd like to see each sass file be directly associated with (and imported by) exactly one react component. then all sass would be directly related to some js view code. If we can make any steps towards making that a reality in this PR let's do it. but we certainly won't get that job done as part of this, because it'll be big
